src/main/cpp/src/SparkResourceAdaptorJni.cpp

```diff
@@ -52,121 +225,832 @@ class spark_resource_adaptor final : public rmm::mr::device_memory_resource {
+              ss << "internal error expected to only wake up blocked threads " << 
+                thread_id_to_wake << " " << as_str(thread->second.state);
+              throw std::runtime_error(ss.str());
+            }
         }
+      }
+    }
+  }
 
-        if (thread->second.split_and_retry_oom_injected > 0) {
-          thread->second.split_and_retry_oom_injected--;
-          throw_java_exception(SPLIT_AND_RETRY_OOM_CLASS, "injected SplitAndRetryOOM");
+  /**
+   * Check to see if any threads need to move to BUFN. This should be
+   * called when a task or shuffle thread becomes blocked so that we can 
+   * check to see if one of them needs to becom BUFN or do a split and rollback.
+   */
+  void check_and_update_for_bufn(std::unique_lock<std::mutex> & lock) {
+    // We want to know if all active tasks have at least one thread that
+    // is effectivley blocked or not.  We could change the definitions here,
+    // but for now this sounds like a good starting point.
+    std::set<long> tasks_with_threads;
+    std::set<long> tasks_with_threads_effectively_blocked;
+    bool is_any_shuffle_thread_blocked = false;
+    // To keep things simple we are going to do multiple passes through
+    // the state. The first is to find out if any shuffle thread is blocked
+    // because if it is, then there is a possability that any task thread
+    // in a shuffle could also be blocked.
+    for (auto thread = threads.begin(); thread != threads.end(); thread++) {
+      switch (thread->second.state) {
+        case SHUFFLE_BLOCKED:
+          is_any_shuffle_thread_blocked = true;
+          break;
+        default:
+          break;
+      }
+    }
+
+    for (auto thread = threads.begin(); thread != threads.end(); thread++) {
+      if (thread->second.task_id >= 0) {
+        tasks_with_threads.insert(thread->second.task_id);
+      }
+
+      switch (thread->second.state) {
+        case TASK_WAIT_ON_SHUFFLE:
+        // fall through
+        case TASK_BUFN_WAIT_ON_SHUFFLE:
+          if (is_any_shuffle_thread_blocked) {
+            tasks_with_threads_effectively_blocked.insert(thread->second.task_id);
+          }
+          break;
+        case TASK_BLOCKED:
+        // fall through
+        case TASK_BUFN_THROW:
+        // fall through
+        case TASK_BUFN_WAIT:
+        // fall through
+        case TASK_BUFN:
+          tasks_with_threads_effectively_blocked.insert(thread->second.task_id);
+          break;
+        default:
+          break;
+      }
+    }
+
+    bool need_to_break_deadlock = tasks_with_threads.size() == tasks_with_threads_effectively_blocked.size();
+    if (need_to_break_deadlock) {
+      // Find the task thread with the lowest priority that is not already BUFN
+      thread_priority to_bufn(-1, -1);
+      bool is_to_bufn_set = false;
+      for (auto thread = threads.begin(); thread != threads.end(); thread++) {
+        switch (thread->second.state) {
+          case TASK_BLOCKED:
+            {
```

Comment:
style nit, shouldn't really need this top level curly bracket.

Reply:
I need the top level curly to prevent `thread_priority current` from escaping that scope and causing issues in other places. I get compile errors without it.

Reply:
Looks like switch/case doesn't affect the scope, so that makes sense!